src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/AccountsReferenceDateController.java

```diff
@@ -0,0 +1,164 @@
+
+            ApiClient apiClient = apiClientService.getApiClient();
+            CompanyProfileApi companyProfile = companyService.getCompanyProfile(companyNumber);
+            SmallFullApi smallFullAccounts = smallFullService.getSmallFullAccounts(apiClient, transactionId, companyAccountsId);
+
+            accountsReferenceDate.setPastDates(getPastValidDates(companyProfile.getAccounts().getNextAccounts().getPeriodEndOn()));
+            accountsReferenceDate.setFutureDates(getFutureValidDates(companyProfile.getAccounts().getNextAccounts().getPeriodEndOn()));
+
+            if(!companyProfile.getAccounts().getNextAccounts().getPeriodEndOn().equals(smallFullAccounts.getNextAccounts().getPeriodEndOn())) {
+                accountsReferenceDate.setChosenDate(smallFullAccounts.getNextAccounts().getPeriodEndOn());
+            }
+
+            model.addAttribute(ACCOUNTS_REFERENCE_DATE, accountsReferenceDate);
+
+            return getTemplateName();
+        } catch (ServiceException e) {
+            LOGGER.errorRequest(request, e.getMessage(), e);
+            return ERROR_VIEW;
+        }
+    }
+
+    @PostMapping
+    public String submitAccountsReferenceDateQuestion(@PathVariable String companyNumber,
+                                                      @PathVariable String transactionId,
+                                                      @PathVariable String companyAccountsId,
+                                                      Model model,
+                                                      @ModelAttribute(ACCOUNTS_REFERENCE_DATE) @Valid AccountsReferenceDate accountsReferenceDate,
+                                                      BindingResult bindingResult,
+                                                      HttpServletRequest request) {
+
+        addBackPageAttributeToModel(model, companyNumber, transactionId, companyAccountsId);
+
+        if (bindingResult.hasErrors()) {
+            return getTemplateName();
+        }
+
+        try {
+            smallFullService.updateSmallFullAccounts(accountsReferenceDate.getChosenDate(), transactionId, companyAccountsId );
+
+            return navigatorService.getNextControllerRedirect(this.getClass(), companyNumber, transactionId, companyAccountsId);
+        } catch (ServiceException e) {
+            LOGGER.errorRequest(request, e.getMessage(), e);
+            return ERROR_VIEW;
+        }
+    }
+
+    @Override
+    protected String getTemplateName() {
+        return "smallfull/accountsReferenceDate";
+    }
+
+    @Override
+    public boolean willRender(String companyNumber, String transactionId, String companyAccountsId)
+            throws ServiceException {
+
+        CompanyAccountsDataState companyAccountsDataState = getStateFromRequest(request);
+        return BooleanUtils.isTrue(!companyAccountsDataState.getHasConfirmedAccountingReferenceDate());
+    }
+
+    private List<LocalDate> getFutureValidDates(LocalDate periodEndOn) {
```

Comment:
Could we maybe abstract this logic into, say, company service? The logic split would make unit testing simpler / more thorough and I think the controller may be doing too much at present - especially when the logic to remove future dates is added.